File: algorithms/[480]滑动窗口中位数/solution.ts
function medianSlidingWindow(nums: number[], k: number): number[] {
  const n = nums.length;
  const arr = nums.slice(0, k);
  arr.sort((a, b) => a - b);

  const res = [findMedian(arr)];

  for (let i = 1; i <= n - k; i += 1) {
    // 删除下标为 i - 1 的元素
    arr.splice(arr.indexOf(nums[i - 1]), 1);
    // 插入下标为 i + k - 1 的元素
    arr.splice(binarySearch(arr, nums[i + k - 1]), 0, nums[i + k - 1]);

    res.push(findMedian(arr));
  }

  return res;
}

function findMedian(nums: number[]): number {
  const n = nums.length;
  return (nums[Math.floor((n - 1) / 2)] + nums[Math.ceil((n - 1) / 2)]) / 2;
}

function binarySearch(nums: number[], target: number): number {
  const n = nums.length;

  let left = 0;
  let right = n;

  while (left < right) {
    const mid = Math.floor((left + right) / 2);
    if (nums[mid] < target) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }

  return left;
}

File: algorithms/[480]滑动窗口中位数/solution.py
# topics = ["滑动窗口", "二分查找", "堆", "二叉搜索树"]

from typing import List
import math
import bisect


class Solution:

    def medianSlidingWindow(self, nums: List[int], k: int) -> List[float]:
        """
        近似题 295.数据流的中位数，此题不同之处在于：需要能（高效）删除任意元素
        二分搜索 + 直接插入
        """
        n = len(nums)
        arr: List[int] = sorted(nums[:k])
        res: List[float] = [self.findMedian(arr, k)]

        for i in range(1, n - k + 1):
            arr.remove(nums[i - 1])
            arr.insert(bisect.bisect(arr, nums[i + k - 1]), nums[i + k - 1])
            res.append(self.findMedian(arr, k))

        return res

    def findMedian(self, nums: List[int], n: int) -> float:
        return (nums[math.floor((n - 1) / 2)] + nums[math.ceil((n - 1) / 2)]) / 2
